apps/shared/avifpng.c

```diff
@@ -162,75 +138,63 @@ static avifBool avifExtractExifAndXMP(png_structp png, png_infop info, avifBool
     png_textp text = NULL;
     const png_uint_32 numTextChunks = png_get_text(png, info, &text, NULL);
     for (png_uint_32 i = 0; (!*ignoreExif || !*ignoreXMP) && (i < numTextChunks); ++i, ++text) {
-        avifRWData metadata = { NULL, 0 };
-        avifBool tagExif00WasRemoved;
+        png_size_t textLength = text->text_length;
+#ifdef PNG_iTXt_SUPPORTED
+        if ((text->compression == PNG_ITXT_COMPRESSION_NONE) || (text->compression == PNG_ITXT_COMPRESSION_zTXt)) {
+            textLength = text->itxt_length;
+        }
+#endif
+
         if (!*ignoreExif && !strcmp(text->key, "Raw profile type exif")) {
-            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
-                return AVIF_FALSE;
-            }
-            uint32_t exifTiffHeaderOffset;
-            if (avifExtractExifTiffHeaderOffset(metadata.data, metadata.size, &exifTiffHeaderOffset) != AVIF_RESULT_OK) {
-                avifRWDataFree(&metadata);
-                fprintf(stderr, "Exif extraction failed: missing TIFF header\n");
+            if (!avifCopyRawProfile(text->text, textLength, &avif->exif)) {
                 return AVIF_FALSE;
             }
-            avifRWDataFree(&avif->exif);
-            avif->exif = metadata;
+            avifRemoveHeadingExif00TagIfAny(&avif->exif);
             *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
         } else if (!*ignoreXMP && !strcmp(text->key, "Raw profile type xmp")) {
-            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+            if (!avifCopyRawProfile(text->text, textLength, &avif->xmp)) {
                 return AVIF_FALSE;
             }
-            if (tagExif00WasRemoved) {
-                avifRWDataFree(&metadata);
-                fprintf(stderr, "XMP extraction failed: XMP payload begins with an Exif tag\n");
-                return AVIF_FALSE;
-            }
-            avifRWDataFree(&avif->xmp);
-            avif->xmp = metadata;
             *ignoreXMP = AVIF_TRUE; // Ignore any other XMP chunk.
         } else if (!strcmp(text->key, "Raw profile type APP1")) {
-            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+            // This can be either Exif, XMP or something else.
+            avifRWData metadata = { NULL, 0 };
+            if (!avifCopyRawProfile(text->text, textLength, &metadata)) {
                 return AVIF_FALSE;
             }
-            // Differentiate Exif and XMP by checking whether the chunk begins with a valid TIFF header ("IIx0x42" or "MMx42x0").
-            const uint32_t tiffHeaderSize = 4;
+            const uint32_t exifTiffHeaderSize = 10; // "MM\0\42" or "II\42\0", maybe prefixed by "Exif\0\0"
```

Comment:
In the raw APP1 case, this MUST be prefixed by `Exif\0\0`, so `exifTiffHeaderSize` is no longer appropriate here (and there's no point in checking for II or MM, you can assume it's a valid payload if the header is matched). Maybe just call this `const exifApp1HeaderSize = 6`?